modules/core/shared/src/main/scala/lucuma/core/optics/ValidSplitEpi.scala

```diff
@@ -0,0 +1,176 @@
+  def unsafeGet(a: A): B =
+    getValid(a).getOrElse {
+      throw new IllegalArgumentException(s"unsafeGet failed: $a")
+    }
+
+  /** Always return a single instance of `E` in case of an invalid `T`. */
+  def withError(e: E): ValidSplitEpi[E, A, B] =
+    ValidSplitEpi(
+      getValid.andThen(_.leftMap(_ => e)),
+      reverseGet
+    )
+
+  /** Build a `ValidWedge` with the same functionality. */
+  def asValidWedge: ValidWedge[E, A, B] =
+    ValidWedge(getValid, reverseGet)
+
+  /** Compose with another `ValidSplitEpi`. */
+  def andThen[C](f: ValidSplitEpi[E, B, C]): ValidSplitEpi[E, A, C] =
+    ValidSplitEpi(
+      getValid(_).fold(_.asLeft, f.getValid),
+      reverseGet.compose(f.reverseGet)
+    )
+
+  /** Compose with a `ValidSplitEpi`. */
+  def andThen[C](f: ValidWedge[E, B, C]): ValidWedge[E, A, C] =
+    asValidWedge.andThen(f)
+
+  /** Compose with a `Format`. */
+  def andThen[C](f: Format[B, C], error: E): ValidSplitEpi[E, A, C] =
+    andThen(ValidSplitEpi.fromFormat(f, error))
+
+  /** Compose with a `Prism`. */
+  def andThen[C](f: Prism[B, C], error: E): ValidSplitEpi[E, A, C] =
+    andThen(ValidSplitEpi.fromPrism(f, error))
+
+  /** Compose with an `Iso`. */
+  def andThen[C](f: Iso[B, C]): ValidSplitEpi[E, A, C] =
+    ValidSplitEpi(
+      getValid(_).map(f.get),
+      reverseGet.compose(f.reverseGet)
+    )
+
+  /** Compose with a `SplitEpi`. */
+  def andThen[C](f: SplitEpi[B, C], error: E): ValidSplitEpi[E, A, C] =
+    andThen(ValidSplitEpi.fromFormat(f.asFormat, error))
+
+  /** Compose with a `SplitMono`. */
+  def andThen[C](f: SplitMono[B, C]): ValidWedge[E, A, C] =
+    ValidWedge(
+      getValid(_).map(f.get),
+      reverseGet.compose(f.reverseGet)
+    )
+
+  /** Compose with a `Wedge`. */
+  def andThen[C](f: Wedge[B, C]): ValidWedge[E, A, C] =
+    ValidWedge(
+      getValid(_).map(f.get),
+      reverseGet.compose(f.reverseGet)
+    )
+
+  /** `ValidSplitEpi` is an invariant functor over A. */
+  def imapA[C](f: A => C, g: C => A): ValidSplitEpi[E, C, B] =
+    ValidSplitEpi(g.andThen(getValid), reverseGet.andThen(f))
+
+  /** `ValidSplitEpi` is an invariant functor over B. */
+  def imapB[C](f: C => B, g: B => C): ValidSplitEpi[E, A, C] =
+    ValidSplitEpi(getValid.andThen(_.map(g)), f.andThen(reverseGet))
+
+  /**
+   * Build `ValidSplitEpi` from another one, but allow empty values to become `None`.
+   */
+  def optional(implicit ev: Monoid[A], eq: Eq[A]): ValidSplitEpi[E, A, Option[B]] =
+    ValidSplitEpi(
+      (a: A) =>
+        if (a.isEmpty)
+          none.asRight
+        else
+          self.getValid(a).map(_.some),
+      (b: Option[B]) => b.foldMap(self.reverseGet)
+    )
+
+  /**
+   * Build `ValidSplitEpi` from another one, refining the return type with predicate `P`.
+   */
+  def refined[P](error: E)(implicit ev: RefinedValidate[B, P]): ValidSplitEpi[E, A, Refined[B, P]] =
+    this.andThen(ValidSplitEpi.forRefined[E, B, P](error))
+}
```

Comment:
Maybe

```
def asFormat: Format[A, B] =
  Format(a => getValid(a).toOption, reverseGet)
```

Reply:
Sorry, this doesn't really have anything to do with the `refined` method.  Just seemed like a handy converter we might add.